### PR TITLE
Explain available copy/cut/paste options better

### LINKED ISSUE #451: can't "copy", "cut", or "paste" via the context menu
For some reason the Brackets context menu has copy/cut/paste disabled (or not even there at all?). That's a big problem for devs, let alone learners =)

### PR DESCRIPTION
**Pull request by mfainshtein4**
implemented a check to see if browser supports these commands via queryCommandSupported if not display dialog which prompts users to use the native functionality for cut, copy, paste. The functionality also includes a check to see if the system is a mac and if so display command+c ..etc and if its not a mac, display Ctrl+c..etc.

fixes #451

### CODE REVIEW

File: src/editor/EditorCommandHandlers.js [hunk 1]
```diff
@@ -1137,6 +1143,38 @@ define(function (require, exports, module) {
...
+	
+	var _execCommandCut = (function() {
+		if(document.queryCommandSupported("cut")) {
+			return function() {
+				_execCommand("cut");
+			};
+		} else {
+			Dialogs.showModalDialogUsingTemplate(dialogTemplate);
+			return ignoreCommand;
+		}
```
Comment (humphd) [inline src/editor/EditorCommandHandlers.js]: This pattern is wrong, here, and also in the code below.  You need to *always* return a function from the anonymous function being run on line 1150.  Basically, you are making `_execCommandCut` a function, but you won't know which one of the two possible options until run time when you check on the value of `document.queryCommandSupported("cut")`.

It should be something like this:

```js
var _execCommandCut = (function() {
    if(document.queryCommandSupported("cut")) {
        return function() {
            _execCommand("cut");
        };
    }

    return function() {
        Dialogs.showModalDialogUsingTemplate(dialogTemplate);
        return ignoreCommand();
    };
}());
```

File: src/editor/EditorCommandHandlers.js [hunk 2]
```diff
@@ -1137,6 +1143,38 @@ define(function (require, exports, module) {
...
+	}());
+	
+	var _execCommandCopy = (function() {
```
Comment (humphd) [inline src/editor/EditorCommandHandlers.js]: Same issue here.

File: src/editor/EditorCommandHandlers.js [hunk 3]
```diff
@@ -1137,6 +1143,38 @@ define(function (require, exports, module) {
...
+			_execCommand("copy");
+		};
+		} else {			
+			Dialogs.showModalDialogUsingTemplate(dialogTemplate);
+			return ignoreCommand;
+		}
+	}());
+	
+	function _execCommandPaste(){
+		Dialogs.showModalDialogUsingTemplate(dialogTemplate);
```
Comment (humphd) [inline src/editor/EditorCommandHandlers.js]: Indent issue.

Also, you need to add `return ignoreCommand();` as well.

File: src/htmlContent/copy-paste-dialog.html [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    <link rel="stylesheet" type="text/css" href="styles/bramble_overrides.css">
+  </head>
+  <div class="modal-content">
+    <span class="dialog-button" data-button-id="escape">&#10006;</span>
+    <h2 class="model-header">Copying and pasting in Thimble</h2>
+    <p class="model-text">
+      These actions are unavailable via the Edit menu, but you can still use:
+    </p>
+    <table class= "guide">
+    <script>
```
Comment (humphd) [inline src/htmlContent/copy-paste-dialog.html]: You don't want to run a script in this dialog.  Instead, you'll want to follow the pattern here:

https://github.com/adobe/brackets/blob/f028b3ecdf2017498196f39ddbbe4a2364366c92/src/help/HelpCommandHandlers.js#L81

Render variables into your template with Mustache.

File: src/htmlContent/copy-paste-dialog.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<div id="myModal" class="modal">
```
Comment (humphd) [inline src/htmlContent/copy-paste-dialog.html]: Can you put a screenshot of this in your PR so @flukeout can review the UI as well?

Comment (mfainshtein4) [inline src/htmlContent/copy-paste-dialog.html]: current
@flukeout it currently looks like this.

File: src/editor/EditorCommandHandlers.js [hunk 6]
```diff
@@ -32,7 +32,7 @@ define(function (require, exports, module) {
         Strings            = require("strings"),
         Editor              = require("editor/Editor").Editor,
         CommandManager     = require("command/CommandManager"),
-		Dialogs        	   = require("widgets/Dialogs"),
+	Dialogs        	   = require("widgets/Dialogs"),
```
Comment (Pomax) [inline src/editor/EditorCommandHandlers.js]: this still uses a tab. Please set your code editor to use spaces for tabs in this file. If you do a text selection on any of the lines above/below you'll see that it's indentation that purely uses spaces, unlike your line here, which uses a tab

**Comment by Pomax** [comment]
@mfainshtein4 can you make sure to run `npm test` before you push up a commit? Right now your code has some linting issues that makes the automated-tester fail, which you can catch by running `npm test` on your own machine and making sure that your code does not fail that, before pushing it up.

**Comment by humphd** [comment]
Looks like a missing semicolon:

```
Running "eslint:src" (eslint) task
/home/travis/build/mozilla/brackets/src/editor/EditorCommandHandlers.js

  1173:10  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
```

@Pomax is right, just like with our lab last week, you can do your testing with `eslint` locally using `npm test` without having to actually push broken commits.  It's good to get things passing locally so you don't waste cycles on travis, or spam people with github notifications for broken builds.

File: src/editor/EditorCommandHandlers.js [hunk 7]
```diff
@@ -45,7 +43,7 @@ define(function (require, exports, module) {
     var commandKey = {
         COMMAND_KEY: "Ctrl"
     };
-    if(true){
+    if(navigator.platform.indexOf("Mac") !== -1;){
         commandKey = {
             COMMAND_KEY: "&#8984"
```
Comment (mfainshtein4) [inline src/editor/EditorCommandHandlers.js]: @humphd This was the code I mentioned in the email. Where the Unicode doesn't display the Command symbol.

Comment (mfainshtein4) [inline src/editor/EditorCommandHandlers.js]: ignore the semicolon I will remove that..

Comment (humphd) [inline src/editor/EditorCommandHandlers.js]: I see in the [Mustache docs](https://github.com/janl/mustache.js#variables) the following: "All variables are HTML-escaped by default. If you want to render unescaped HTML, use the triple mustache: {{{name}}}. You can also use & to unescape a variable."

File: src/htmlContent/copy-paste-dialog.html [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  <div class="modal-content">
+    <span class="dialog-button" data-button-id="escape">&#10006;</span>
+    <h2 class="model-header">Copying and pasting in Thimble</h2>
+    <p class="model-text">
+      These actions are unavailable via the Edit menu, but you can still use:
+    </p>
+    <table class= "guide">
+      <tr>
+        <th>
+          {{COMMAND_KEY}}+X
```
Comment (humphd) [inline src/htmlContent/copy-paste-dialog.html]: Try making these `{{{COMMAND_KEY}}}` (3 mustaches).

File: src/styles/bramble_overrides.css [hunk 9]
```diff
@@ -72,6 +72,29 @@
     background: rgba(0, 135, 207, .3) !important;
 }
 
+table.guide {
+	width: 100%;
```
Comment (humphd) [inline src/styles/bramble_overrides.css]: Your indents are all too far in this file.  4-space indent please, match to surrounding code.

File: src/command/DefaultMenus.js [hunk 10]
```diff
@@ -253,11 +253,11 @@ define(function (require, exports, module) {
         // editor_cmenu.addMenuItem(Commands.NAVIGATE_JUMPTO_DEFINITION);
         editor_cmenu.addMenuItem(Commands.TOGGLE_QUICK_EDIT);
         editor_cmenu.addMenuItem(Commands.TOGGLE_QUICK_DOCS);
-        editor_cmenu.addMenuDivider();
-        editor_cmenu.addMenuItem(Commands.EDIT_CUT);
-        editor_cmenu.addMenuItem(Commands.EDIT_COPY);
-        editor_cmenu.addMenuItem(Commands.EDIT_PASTE);
-        editor_cmenu.addMenuDivider();
+		editor_cmenu.addMenuDivider();
```
Comment (Pomax) [inline src/command/DefaultMenus.js]: Please use the correct indentation, and please make sure to run `npm test` before committing code, because `npm test` will inform you of these easy to catch errors. They should not show up in a pushed commit

File: src/editor/EditorCommandHandlers.js [hunk 11]
```diff
@@ -1142,26 +1148,53 @@ define(function (require, exports, module) {
 
         return result.promise();
     }
+<<<<<<< HEAD
```
Comment (Pomax) [inline src/editor/EditorCommandHandlers.js]: this is a file with merge conflict code in. Please fix your commits first, and verify that your `npm test` *and* `npm run build` work. There is no way this file works.

File: src/command/DefaultMenus.js [hunk 12]
```diff
@@ -253,6 +253,11 @@ define(function (require, exports, module) {
         // editor_cmenu.addMenuItem(Commands.NAVIGATE_JUMPTO_DEFINITION);
         editor_cmenu.addMenuItem(Commands.TOGGLE_QUICK_EDIT);
         editor_cmenu.addMenuItem(Commands.TOGGLE_QUICK_DOCS);
+		editor_cmenu.addMenuDivider();
```
Comment (Pomax) [inline src/command/DefaultMenus.js]: please run `npm test` before you push changes, as that will tell you that your indentation here needs fixing.

File: src/editor/EditorCommandHandlers.js [hunk 13]
```diff
@@ -1137,6 +1148,41 @@ define(function (require, exports, module) {
         return result.promise();
     }
 
+    function _execCommand(cmd) {
+        window.document.execCommand(cmd);
+	}
```
Comment (Pomax) [inline src/editor/EditorCommandHandlers.js]: indentation

File: src/command/DefaultMenus.js (hunk 12 above)
Comment (mfainshtein4) [inline src/command/DefaultMenus.js]: when I run npm test it doesn't say anything is wrong? why could this be?

Comment (humphd) [inline src/command/DefaultMenus.js]: I don't think the Brackets code does much in the way of lint-format checks.

File: src/editor/EditorCommandHandlers.js [hunk 14]
```diff
@@ -30,14 +30,25 @@ define(function (require, exports, module) {
...
         StringUtils        = require("utils/StringUtils"),
         TokenUtils         = require("utils/TokenUtils"),
         CodeMirror         = require("thirdparty/CodeMirror/lib/codemirror"),
+        Dialogs            = require("widgets/Dialogs"),
+        dialogTemplate     = require("text!htmlContent/copy-paste-dialog.html"),
+        Mustache           = require("thirdparty/mustache/mustache"),
         _                  = require("thirdparty/lodash");
-
+	
+    var commandKey = {
```
Comment (humphd) [inline src/editor/EditorCommandHandlers.js]:
```js
var commandKey;
if(brackets.platform === "mac") {
    commandKey = ...;
} else {
    commandKey = ...;
}
```

File: src/editor/EditorCommandHandlers.js [hunk 15]
```diff
@@ -1137,6 +1148,41 @@ define(function (require, exports, module) {
...
+                _execCommand("copy");
+            };
+        }
+        return function() {
+            Dialogs.showModalDialogUsingTemplate(Mustache.render(dialogTemplate, commandKey));
+            return ignoreCommand();
+        };
+    }());
+	
+    function _execCommandPaste(){
```
Comment (humphd) [inline src/editor/EditorCommandHandlers.js]: Add a comment above this indicating that we ignore the value of `document.queryCommandSupported("paste")`, because browsers don't honour it due to security concerns.

File: src/htmlContent/copy-paste-dialog.html (hunk 5 above)
Comment (humphd) [inline src/htmlContent/copy-paste-dialog.html]: @flukeout we need a UI review from you on this change.

File: src/styles/bramble_overrides.css [hunk 16]
```diff
@@ -86,13 +111,16 @@
...
 }
 
-
 /* Makes the resize handles wider, and visile on hover & drag */
 
 #sidebar {
   position: relative;
 }
 
+.CodeMirror-foldgutter:after {
```
Comment (humphd) [inline src/styles/bramble_overrides.css]: Why this change?

Comment (mfainshtein4) [inline src/styles/bramble_overrides.css]: During a merge conflict I just appended it under my stuff. I wasn't sure what its used for. So I kept it there.

**Comment by flukeout** [comment]
Testing this out now.

**Comment by flukeout** [comment]
@mfainshtein4 how do I get this UI to show up in my local environment? I guess my browsers support the operation, but is there something I can change to preview the popup? Let me know, thanks!

**Comment by flukeout** [comment]
Nevermind @mfainshtein4 - I got it to show up when I use the "Paste" action. However, on my mac, it looks like the key type (Command) is missing from the instructions...

image

**Comment by humphd** [comment]
@mfainshtein4 is this ready for review again?  When it is, use @ and our names to make sure we see this.

**Comment by mfainshtein4** [comment]
@Pomax , @humphd , @flukeout I am ready for another revision.

**Comment by flukeout** [comment]
Checking in on the latest changes now.

**Comment by flukeout** [comment]
@humphd Please merge it! I have some minor suggestions but I'll make those as a separate PR.

@mfainshtein4 - thank you so much for this work! 👍

**Comment by humphd** [comment]
A long road, but we finally made it.  I've tested locally, and this is working great.  Thanks for sticking it out, @mfainshtein4, this is a good fix.

**Comment by mfainshtein4** [comment]
@humphd Thanks. I appreciate the constant support and pokes in the right direction!
